earthfile2llb/wait_block.go

```diff
@@ -0,0 +1,194 @@
+package earthfile2llb
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"sync"
+
+	"github.com/earthly/earthly/states"
+	"github.com/earthly/earthly/util/llbutil"
+	"github.com/earthly/earthly/util/llbutil/pllb"
+	"github.com/moby/buildkit/exporter/containerimage/exptypes"
+	"github.com/moby/buildkit/frontend/gateway/client"
+	gwclient "github.com/moby/buildkit/frontend/gateway/client"
+	"github.com/pkg/errors"
+)
+
+type saveImagesWaitItem struct {
+	converters []*Converter
+	images     []states.SaveImage
+	push       []bool
+}
+
+type runCommandWaitItem struct {
+	c   *Converter
+	cmd *pllb.State
+}
+
+type waitItem interface {
+	wait(context.Context) error
+}
+
+func (w *saveImagesWaitItem) wait(ctx context.Context) error {
+	isMultiPlatform := make(map[string]bool)    // DockerTag -> bool
+	noManifestListImgs := make(map[string]bool) // DockerTag -> bool
+	platformImgNames := make(map[string]bool)
+
+	for _, saveImage := range w.images {
+		if saveImage.NoManifestList {
+			noManifestListImgs[saveImage.DockerTag] = true
+		} else {
+			isMultiPlatform[saveImage.DockerTag] = true
+		}
+		if isMultiPlatform[saveImage.DockerTag] && noManifestListImgs[saveImage.DockerTag] {
+			return fmt.Errorf("cannot save image %s defined multiple times, but declared as SAVE IMAGE --no-manifest-list", saveImage.DockerTag)
+		}
+	}
+
+	metadata := map[string][]byte{}
+	refs := map[string]client.Reference{}
+
+	refID := 0
+	for imageIndex, saveImage := range w.images {
+		if !w.push[imageIndex] {
+			continue
+		}
+		c := w.converters[imageIndex]
+
+		ref, err := llbutil.StateToRef(
+			ctx, c.opt.GwClient, saveImage.State, c.opt.NoCache,
+			c.platr, c.opt.CacheImports.AsMap())
+		if err != nil {
+			return errors.Wrapf(err, "failed to solve image required for %s", saveImage.DockerTag)
+		}
+
+		config, err := json.Marshal(saveImage.Image)
+		if err != nil {
+			return errors.Wrapf(err, "marshal save image config")
+		}
+
+		refKey := fmt.Sprintf("image-%d", refID)
+		refPrefix := fmt.Sprintf("ref/%s", refKey)
+		refs[refKey] = ref
+
+		metadata[refPrefix+"/image.name"] = []byte(saveImage.DockerTag)
+		metadata[refPrefix+"/export-image-push"] = []byte("true")
+		if saveImage.InsecurePush {
+			metadata[refPrefix+"/insecure-push"] = []byte("true")
+		}
+		metadata[refPrefix+"/"+exptypes.ExporterImageConfigKey] = config
+		refID++
+
+		if isMultiPlatform[saveImage.DockerTag] {
+			platformStr := saveImage.Platform.String()
+			platformImgName, err := llbutil.PlatformSpecificImageName(saveImage.DockerTag, saveImage.Platform)
+			if err != nil {
+				return err
+			}
+
+			if saveImage.CheckDuplicate && saveImage.DockerTag != "" {
+				if _, found := platformImgNames[platformImgName]; found {
+					return errors.Errorf(
+						"image %s is defined multiple times for the same platform (%s)",
+						saveImage.DockerTag, platformImgName)
+				}
+				platformImgNames[platformImgName] = true
+			}
+
+			metadata[refPrefix+"/platform"] = []byte(platformStr)
+		}
+	}
```

Comment:
The above should be abstracted somehow so that we reuse the code in both the builder and here... Too much brittle duplication.

Reply:
What are your thoughts on 6d0cc0e89913036406d0d9d4f68d7a4927e90b5a ? It manages to only define the applicable metadata tags in a single location.

Perhaps we could expand this helper for additional earthlyoutput exporter options in a follow up PR?

Reply:
Yeah, sounds good!